### PR TITLE
agent: enhance get handled signal

### PR DESCRIPTION
**Pull request by quanweiZhou**
Get handled signal include SigBlk, SigIgn and SigCgt.

Fixes: #4478
Signed-off-by: quanweiZhou <quanweiZhou@linux.alibaba.com>

### CODE REVIEW

**Comment by quanweiZhou** [comment]
> > Get handled signal include SigBlk, SigIgn and SigCgt.
> > Fixes: #4478
> > Signed-off-by: quanweiZhou [quanweiZhou@linux.alibaba.com](mailto:quanweiZhou@linux.alibaba.com)
> 
> @quanweiZhou Please provide details in the issue on what you're trying to achieve with this patch.

Thanks @gkurz update the comment

**Comment by c3d** [comment]
> Thanks @gkurz, the comment has been update.

@quanweiZhou I do not see the updated comment in the new commit message.

Essentially, I think that the commit message needs to explain, at lesat to the level described in issue #4478 : 
* Why is it a problem to turn `SIGTERM` into `SIGKILL` (I think, based on [this comment](https://github.com/kata-containers/kata-containers/issues/4478#issuecomment-1165313855) that your rationale is that we should leave that policy to Kubernetes, which does it on its own)
* Why we should send a signal to a process that ignores it. Blocking is often temporary, so it could make sense to not immediately convert to a `SIGKILL`, if if the process is ignored, you need to make it clear that you send the process to be ignored only because you expect Kubernetes to later request a `SIGKILL`.
* What about the non-Kubernetes cases? To what extent do we rely on a policy of the orchestration layer that is not necessarily guaranteed?

All this makes me wonder if the simplest would not be to simply pass the signals we receive as is. Does it make any sense to turn `SIGTERM` into `SIGKILL`? If so, why does it make sense for Kata and not for `runc` or `crun`?

File: src/agent/src/rpc.rs [hunk 1]
```diff
@@ -1802,7 +1802,10 @@ fn is_signal_handled(proc_status_file: &str, signum: u32) -> bool {
...
             }
         };
-        if line.starts_with("SigCgt:") {
+
+        // check SigBlk, SigIgn, SigCgt
+        if line.starts_with("SigBlk:") || line.starts_with("SigIgn:") || line.starts_with("SigCgt:")
+        {
             let mask_vec: Vec<&str> = line.split(':').collect();
             if mask_vec.len() != 2 {
                 warn!(sl!(), "parse the SigCgt field failed");
```
Comment (Tim-Zhang) [inline src/agent/src/rpc.rs:1811]: It seems that the loop should continue to try the next sigmask

**Comment by quanweiZhou** [comment]
> > Thanks @gkurz, the comment has been update.
> 
> @quanweiZhou I do not see the updated comment in the new commit message.
> 
> Essentially, I think that the commit message needs to explain, at lesat to the level described in issue #4478 :
> 
> * Why is it a problem to turn `SIGTERM` into `SIGKILL` (I think, based on [this comment](https://github.com/kata-containers/kata-containers/issues/4478#issuecomment-1165313855) that your rationale is that we should leave that policy to Kubernetes, which does it on its own)
> * Why we should send a signal to a process that ignores it. Blocking is often temporary, so it could make sense to not immediately convert to a `SIGKILL`, if if the process is ignored, you need to make it clear that you send the process to be ignored only because you expect Kubernetes to later request a `SIGKILL`.
> * What about the non-Kubernetes cases? To what extent do we rely on a policy of the orchestration layer that is not necessarily guaranteed?
> 
> All this makes me wonder if the simplest would not be to simply pass the signals we receive as is. Does it make any sense to turn `SIGTERM` into `SIGKILL`? If so, why does it make sense for Kata and not for `runc` or `crun`?

Thanks @c3d update the comment
```
For runC, send the signal to the init process directly.
For kata, we try to send `SIGKILL` instead of `SIGTERM` when the process has not installed the handler for `SIGTERM`.
The `is_signal_handled` function determine which signal the container process has been handled. But currently `is_signal_handled` is only catching (SigCgt). While the container process is ignoring (SigIgn) or blocking (SigBlk) also should not be converted from the `SIGTERM` to `SIGKILL`. For example, when using terminationGracePeriodSeconds the k8s will send SIGTERM first and then send `SIGKILL`, in this case, the container ignores the `SIGTERM`, so we should send the `SIGTERM` not the `SIGKILL` to the container.
```

File: src/agent/src/rpc.rs (hunk 1 above)
Comment (quanweiZhou) [reply src/agent/src/rpc.rs]: Thanks @Tim-Zhang, change to filter to check signal mask.

**Review by Tim-Zhang — APPROVED** [review]
LGTM thanks @quanweiZhou

**Comment by bergwolf** [comment]
@c3d 
> All this makes me wonder if the simplest would not be to simply pass the signals we receive as is. Does it make any sense to turn SIGTERM into SIGKILL? If so, why does it make sense for Kata and not for runc or crun?

To further explain this one, for some init containers, they may not install the SIGTERM handler. If we do not turn SIGTERM into SIGKILL in such case, containerd would still send SIGKILL later on, after a long delay, which is not what we want when stopping the pod. See the comments in the `do_signal_process` function.

**Review by bergwolf — APPROVED** [review]

**Pull request merged**